src/nvim/ex_docmd.c

```diff
@@ -216,6 +216,38 @@ static void restore_dbg_stuff(struct dbg_stuff *dsp)
   current_exception = dsp->current_exception;
 }
 
+/// Check if ffname differs from fnum.
+/// fnum is a buffer number. 0 == current buffer, 1-or-more must be a valid buffer ID.
+/// ffname is a full path to where a buffer lives on-disk or would live on-disk.
```

Comment:
Should it check that ffname is actually a full path (or expand it)? Is `eap->arg` in `ex_edit` guaranteed to be a full path?

Reply:
I'm not at my machine so I can't give you hard data but I remember ffname wasn't a silver bullet

These were the considerations that I had to make while writing the code, generally

- Does it live on disk
- What is the user's current `:pwd`
    - Are they referring to a buffer with a relative path that does or does not exist on-disk (if relative, what's the `:pwd`)
- Does `curbuf->fname` point to `curbuf->ffname` or `curbuf->sfname` and is `curbuf->ffname` the same as `curbuf->sname` currently

In the end I actually got all of the weirdest edge cases working in Neovim but I used a variable that Vim doesn't guarantee to exist. So I haven't submitted a PR yet. Luckily the unsolved cases are pretty edge-of-edge case so I figured we could take our time with solving them

Reply:
Yeah this is a pain [for plugins](https://github.com/justinmk/vim-dirvish/blob/b660af1fa07fe1d44d4eb3ea5242334f6c2766ca/autoload/dirvish.vim#L459-L470) too.

If there isn't already, we might want to move this function to a more re-usable place, for internal use at least. Externally, `vim.fs.normalize()` is supposed to help with comparing filepaths, but still doesn't cover the weird edge cases of Vim buffer equivalence, so we might need e.g. a `bufequal()` function.

Reply:
That does sound like a `bufequal` function would be pretty useful. Probably worth mentioning an effort to standardize would likely run into https://github.com/vim/vim/pull/14286#issuecomment-2023134377 unless a workaround that works for Vim and Neovim is found. But FWIW I have all of that code working in Neovim so we could implement it whenever we like. Want me to make a `bufequal` GitHub issue on https://github.com/vim/vim or should we hold off?

Reply:
> Want me to make a `bufequal` GitHub issue on https://github.com/vim/vim

SGTM!

Reply:
Added here: https://github.com/vim/vim/issues/14340